data-access-layer/iaas/AliClient.js

```diff
@@ -0,0 +1,212 @@
+const Storage = require('ali-oss');
+// const Compute = require('@alicloud/pop-core');
+const logger = require('../../common/logger');
+// const errors = require('../../common/errors');
+// const utils = require('../../common/utils');
+// const uuid = require('uuid');
+// const ComputeClient = require('./ComputeClient');
+const BaseCloudClient = require('./BaseCloudClient');
+// const NotFound = errors.NotFound;
+// const Unauthorized = errors.Unauthorized;
+// const Forbidden = errors.Forbidden;
+
+class AliClient extends BaseCloudClient {
+  constructor(settings) {
+    super(settings);
+    // this.constructor.validateParams(_.chain(this.settings).value());
+    this.storage = this.constructor.createStorageClient(_
+      .chain(this.settings)
+      .omit('name')
+      .set('provider', this.provider)
+      .value()
+    );
+  }
+
+  getContainer(container) {
+    if (arguments.length < 1) {
+      container = this.containerName;
+    }
+    logger.debug('Looking for container ' + container);
+    return Promise.try(() => {
+      return this.storage.listBuckets({
+        prefix: container
+      })
+        .then(buckets => {
+          if (buckets.buckets === null) {
+            logger.error('Bucket ' + container + ' does not exists');
+          } else if (buckets.buckets.length == 1) {
+            return buckets.buckets;
+            logger.info('Bucket ' + container + ' exists');
+          } else {
+            logger.error('More than 1 Buckets with prefix ' + container + ' exists');
+          }
+        })
+        .catch(err => {
+          logger.error('Bucket ' + container + ' does not exists');
+        });
+    });
+  }
+
+
+  list(container, options) {
+    if (arguments.length < 2) {
+      options = container;
+      container = this.containerName;
+    }
+    return this.storage
+      .useBucket(container)
+      .list(options)
+      .then(listOfFiles => {
+        let list = listOfFiles.objects;
+        let isTruncated = listOfFiles.isTruncated;
+        let marker = listOfFiles.nextMarker;
+        const files = [];
+        _.each(list, file => files.push(_
+          .chain(file)
+          .pick('name', 'lastModified')
+          .set('isTruncated', isTruncated)
+          .set('marker', marker)
+          .value()
+        ));
+        return files;
+      });
+  }
+
+  listFilenames(prefix, max_keys) {
+    const options = {
+      prefix: prefix,
+      'max-keys': max_keys
+    };
+
+    let fileList = [];
+    let level = 0;
+
+    function fetchFiles() {
+      level++;
+      logger.debug(`Fetching recursively at level : ${level}`);
+      const promise = new Promise(function (resolve, reject) {
+        Promise.try(() => list(this.containerName, options))
+          .then(files => {
+            logger.debug('list of files recieved - ', files);
+            if (files && files.length > 0) {
+              fileList = fileList.concat(files);
+              if (files[0].isTruncated === true && level < 10) {
```

Comment:
Please check if this pagination logic is valid for Alicloud.

Reply:
checked and verified that it works for Ali.